gson/src/main/java/com/google/gson/internal/bind/ObjectTypeAdapter.java

```diff
@@ -76,7 +76,7 @@
       return in.nextString();
 
     case NUMBER:
-      return in.nextDouble();
+      return in.nextNumber();
```

Comment:
This might be backward incompatible in case existing code expects a `Double` instead of any `Number`.